### PR TITLE
Block instrumentation improvements

### PR DESCRIPTION
**Pull request by tmat**
Simplifies block instrumentation and makes it composable (allows multiple instrumenters to add local variables).

Fixes sequence point applied to instrumentation prologue.

### CODE REVIEW

**Comment by tmat** [comment]
@jcouv @jjonescz @dotnet/roslyn-compiler ptal

File: src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs [hunk 1]
```diff
@@ -734,6 +735,29 @@ internal static ImmutableArray<T> Concat<T>(this ImmutableArray<T> first, T seco
...
+                return self.Add(items[0]);
+            }
+
+            var builder = ArrayBuilder<T>.GetInstance(self.Length + items.Count);
+            builder.AddRange(self);
+
+            foreach (var item in items)
+            {
+                builder.Add(item);
+            }
```
Comment (jjonescz) [inline src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs:756]:
```suggestion
            builder.AddRange(items);
```

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs [hunk 2]
```diff
@@ -580,6 +586,7 @@ public override BoundNode VisitTypeOrInstanceInitializers(BoundTypeOrInstanceIni
                     if (initializer.Kind == BoundKind.Block)
                     {
                         var block = (BoundBlock)initializer;
+
```
Comment (jjonescz) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs:589]:
```suggestion
```

**Review by jjonescz — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Lowering/Instrumentation/DebugInfoInjector.cs [hunk 3]
```diff
@@ -140,39 +142,41 @@ public override BoundStatement InstrumentYieldReturnStatement(BoundYieldReturnSt
...
+            else if (original == rewriter.CurrentMethodBody)
+            {
+                if (previousPrologue != null)
+                {
+                    prologue = BoundSequencePoint.CreateHidden(previousPrologue);
+                }
+                else if (rewriter.Factory.TopLevelMethod is SynthesizedSimpleProgramEntryPointSymbol)
+                {
+                    prologue = BoundSequencePoint.CreateHidden();
+                }
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/Instrumentation/DebugInfoInjector.cs:173]: What is the impact of this addition? Does this fix a bug?

Comment (cston) [reply src/Compilers/CSharp/Portable/Lowering/Instrumentation/DebugInfoInjector.cs:173]: Is this just a move of code from `LocalRewriter.VisitBlock()`?

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Lowering/Instrumentation/DebugInfoInjector.cs (hunk 3 above)
Comment (tmat) [reply src/Compilers/CSharp/Portable/Lowering/Instrumentation/DebugInfoInjector.cs:173]: Yes, just move.

File: src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs (hunk 1 above)
Comment (tmat) [reply src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs:756]: The builder does not have an overload that takes `TemporaryArray<T>`

**Pull request merged**